[TIMOB-6392] Android fix: Ti.UI.Button text values with too many dis…

maggieaxway opened the pull request:
…playable characters draw improperly

[TIMOB-6392]- Android fix: Ti.UI.Button text values with too many
displayable characters draw improperly

**JIRA:** https://jira.appcelerator.org/browse/TIMOB-6392

Test case:

````js


var window = Ti.UI.createWindow({ layout: "vertical" });
window.add(Ti.UI.createButton(
{
	title: "Single Line Button",
	top: "5%",
}));
window.add(Ti.UI.createButton(
{
	title: "Multiline\nButton",
	top: "5%",
}));
window.add(Ti.UI.createButton(
{
	title: "Single Line Truncated Button",
	top: "5%",
	width: "50%",
}));
window.add(Ti.UI.createButton(
{
	title: "Multiline Button\nThis is the next very long line.",
	top: "5%",
	width: "50%",
}));
window.add(Ti.UI.createButton(
{
	title: "Very long multiline button with word wrapping.",
	top: "5%",
	width: "50%",
}));
window.add(Ti.UI.createButton(
{
	title: "Very long Multiline button with word wrapping.",
	top: "5%",
	width: "50%",
	height: "5%",
}));
window.open();

maggieaxway committed:
[TIMOB-6392]- Android fix: Ti.UI.Button text values with too many displayable characters draw improperly

[TIMOB-6392]- Android fix: Ti.UI.Button text values with too many
displayable characters draw improperly

jquick-axway commented:
@maggieaxway, I did a quick test on iOS and it turn out we don't support multiline buttons on that platform. That is, there is no line wrapping and it ignores newline '\n' characters. It's always forced to be single-line. iOS also always applies ellipsis in the middle of the text as well.

So, for parity with iOS, let's do the following in the TiUIButton constructor:
- Force to single line via `setMaxLines(1)`.
- Use `TextUtils.TruncateAt.MIDDLE` ellipsis mode instead. (Matches iOS behavior.)

This also means you don't need to do the above dynamically anymore. So, you can get rid of the `onGlobalLayout()` listener. This is also a breaking change and should only go into Titanium 7.0 (don't back-port it), meaning that multiline button support will be lost. But it's better to add new button properties to both Android and iOS to enable such a feature to make it portable.

maggieaxway committed:
Update TiUIButton.java

For parity with iOS:

Force to set single line
Use TextUtils.TruncateAt.MIDDLE ellipsis mode

jquick-axway commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButton.java:
```diff
@@ -84,17 +83,8 @@ public void processProperties(KrollDict d)
-				}
-			});
+			btn.setEllipsize(TextUtils.TruncateAt.MIDDLE);
+			btn.setMaxLines(1);
```
Can you move these 2 lines into the TiUIButton constructor please?
(Before the `setNativeView(btn)` call.)

This way these setting will always be applied. This is in case the "title" property is not set upon button creation and is set afterwards instead.

Other than that, yeah, it should be this simple. :)

maggieaxway replied:
Yes you're right! I didn't notice it. Appreciate your caution!

maggieaxway committed:
Update TiUIButton.java

Move changes to constructor to make sure always applied

jquick-axway reviewed (approved):
CR: Pass

ssjsamir committed:
Merge branch 'master' into Timob-6392

ssjsamir committed:
Merge branch 'master' into Timob-6392

lokeshchdhry commented:
FR Passed.

For longer button text on android the text will be shortened using ellipses like IOS.

Studio Ver:     4.10.0.201709271713
SDK Ver:        7.0.0 local build
OS Ver:         10.12.3
Xcode Ver:      Xcode 8.3.3
Appc NPM:       4.2.10
Appc CLI:       6.3.0
Ti CLI Ver:     5.0.14
Alloy Ver:      1.10.7
Node Ver:       7.10.1
Java Ver:       1.8.0_101
Devices:        ⇨ google Nexus 6P --- Android 8.0.0
                ⇨ google Nexus 5 --- Android 6.0.1

Commit:
Merge branch 'master' into Timob-6392

build commented:
<!--
  1 failure:  :microscope: Ther...
  0 warning: 
  
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/maggieaxway/titanium_mobile/blob/Timob-6392/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

lokeshchdhry commented:
@maggieaxway , Jenkins is asking for unit tests before merge. Can you please add it.

jquick-axway commented:
I don't think there is a good way to automate testing of this particular case. You have to eyeball the text and make sure it's showing ellipsis.

maggieaxway commented:
@lokeshchdhry There's no crash before or after the changes, it just meant to display properly. So I agree with Josh it doesn't make sense to make auto test.

lokeshchdhry commented:
@sgtcoolguy , Can you please merge this as according to @maggieaxway no unit tests can be added. Thanks!
